jormungandr/src/explorer/graphql/mod.rs

```diff
@@ -255,72 +613,75 @@ impl From<blockcfg::BlockDate> for BlockDate {
-            id,
-            block_hash: Some(block_hash),
-            contents: None,
-        })
+        if block_hashes.is_empty() {
+            return Err(ErrorKind::NotFound(format!("transaction not found: {}", &id,)).into());
+        } else {
+            Ok(Transaction {
+                id,
+                block_hashes,
+                contents: None,
+            })
+        }
     }
 
     fn from_valid_id(id: FragmentId) -> Transaction {
         Transaction {
             id,
-            block_hash: None,
+            block_hashes: Default::default(),
             contents: None,
         }
     }
 
     fn from_contents(contents: ExplorerTransaction) -> Transaction {
         Transaction {
             id: contents.id,
-            block_hash: None,
+            block_hashes: Default::default(),
             contents: Some(contents),
         }
     }
 
-    async fn get_block(&self, context: &Context) -> FieldResult<Arc<ExplorerBlock>> {
-        let block_id = match self.block_hash {
-            Some(block_id) => block_id,
-            None => context
-                .db
-                .get_main_tip()
-                .await
-                .1
-                .state()
-                .find_block_hash_by_transaction(&self.id)
-                .ok_or_else(|| {
-                    ErrorKind::InternalError("Transaction's block was not found".to_owned())
-                })?,
+    async fn get_blocks(&self, context: &Context) -> FieldResult<Vec<Arc<ExplorerBlock>>> {
+        let block_ids = if self.block_hashes.is_empty() {
+            context.db.find_blocks_by_transaction(&self.id).await
+        } else {
+            self.block_hashes.clone()
         };
 
-        context.db.get_block(&block_id).await.ok_or_else(|| {
-            ErrorKind::InternalError(
-                "transaction is in explorer but couldn't find its block".to_owned(),
-            )
-            .into()
-        })
+        if block_ids.is_empty() {
+            return Err(FieldError::from(ErrorKind::InternalError(
+                "Transaction is not present in any block".to_owned(),
+            )));
+        }
+
+        let mut result = Vec::new();
+
+        for block_id in block_ids {
+            let block = context.db.get_block(&block_id).await.ok_or_else(|| {
+                FieldError::from(ErrorKind::InternalError(
+                    "transaction is in explorer but couldn't find its block".to_owned(),
+                ))
+            })?;
+
+            result.push(block);
+        }
+
+        Ok(result)
     }
 
     async fn get_contents(&self, context: &Context) -> FieldResult<ExplorerTransaction> {
         if let Some(c) = &self.contents {
             Ok(c.clone())
         } else {
-            let block = self.get_block(context).await?;
+            //TODO: maybe store transactions outside blocks? as Arc, as doing it this way is pretty wasty
+            let block = &self.get_blocks(context).await?[0];
```

Comment:
Can there be a `get_block` operation if you already know you can use a block with the specific hash?

Reply:
I'm not sure if I understand. There will always be at least a block there, you mean a function that gets that one?

To be honest, this query is not designed really well, getting the block to get the transaction body is kinda bad, because it also means transactions are duplicated if they are in multiple blocks, which is kinda a waste of memory.

Fetching the entire block to get only one transaction it's probably no ideal either.

I didn't want to beat around the bush too much in this PR, so I just left what has equivalent behavior to the previous thing. (Also, extracting transaction bodies from the blocks is not that trivial anyway, because of gc and other things).

Reply:
I mean it's already wasteful to retrieve several blocks if you only need one, and you can just pick the first of `self.block_hashes` and query just that block by its hash.

Optimizing to be able to query individual fragments straight from the store/db would be good (and might be possible with some zero-copy block storage access and iteration on raw fragments in a raw block), but this is a topic for a more substantial rework that would involve the storage layer, so not for this PR. Cc @eugene-babichenko

Reply:
oh, that... Yeah, sure, makes sense.

Well, considering that the explorer doesn't even use the storage, yeah, that's a big change. I'm not sure we want to do that, though? And using raw fragments wouldn't play that nice with the way we show utxo inputs now, I guess.